test/dynamo/test_activation_checkpointing.py

```diff
@@ -1994,6 +1999,238 @@ def forward(self, primals_1: "f32[4, 4]"):
+            torch.utils.checkpoint.create_selective_checkpoint_contexts, policy_fn
+        )
+
+        def fwd_bwd_with_policy(x, w1, b1):
+            def checkpoint_fn(inp, w, b):
+                linear = torch.nn.functional.linear(inp, w, b)
+                return torch.relu(linear)
+
+            result = torch.utils.checkpoint.checkpoint(
+                checkpoint_fn, x, w1, b1, use_reentrant=False, context_fn=context_fn
+            )
+            loss = result.sum()
+
+            with torch.fx.traceback.annotate({"remat_pass_tag": "is_backward"}):
+                dx, dw, db = _grad(loss, (x, w1, b1))
+            return dx, dw, db
+
+        result_with, gm_with = self._compile_and_capture(
+            fwd_bwd_with_policy, True, (x, w1, b1)
+        )
+        result_without, gm_without = self._compile_and_capture(
+            fwd_bwd_with_policy, False, (x, w1, b1)
+        )
+
+        torch.testing.assert_close(result_with[0], result_without[0])
+        torch.testing.assert_close(result_with[1], result_without[1])
+        torch.testing.assert_close(result_with[2], result_without[2])
+
+        addmm_without = self.count_op(gm_without, torch.ops.aten.addmm.default)
+        relu_without = self.count_op(gm_without, torch.ops.aten.relu.default)
+
+        addmm_with = self.count_op(gm_with, torch.ops.aten.addmm.default)
+        relu_with = self.count_op(gm_with, torch.ops.aten.relu.default)
+
+        self.assertEqual(addmm_without, addmm_with)
+        self.assertEqual(relu_with, relu_without + 1)
+
+        recomputed_nodes = [
+            n.name for n in gm_with.graph.nodes if "_recomputed" in n.name
+        ]
+        self.assertNotIn("addmm_recomputed", recomputed_nodes)
+
+        self.assertTrue(
+            any("relu" in name for name in recomputed_nodes),
+            f"Expected relu_recomputed but got: {recomputed_nodes}",
+        )
+
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
+    def test_ac_rematerialize_transitive_dependency_sorting(self):
+        x = torch.randn(4, 4, requires_grad=True)
+        y = torch.randn(4, 4, requires_grad=True)
+        z = torch.randn(4, 4, requires_grad=True)
+
+        def fwd_bwd_with_transitive_deps(x, y, z):
+            def _util_checkpoint(x, y, z):
+                a = x.clone()
+                b = y.clone()
+                c = a + b
+                d = z.clone()
+                return c, d
+
+            c, d = torch.utils.checkpoint.checkpoint(
+                _util_checkpoint, x, y, z, use_reentrant=False
+            )
+
+            with torch.fx.traceback.annotate({"remat_pass_tag": "is_backward"}):
+                e = d + c
+                loss = e.sum()
+                dx = _grad(loss, x)[0]
+
+            return dx.detach()
+
+        _, captured_gm = self._compile_and_capture(
+            fwd_bwd_with_transitive_deps, True, (x, y, z)
+        )
+
+        # If we don't sort, we see following sequence:
```

Comment:
You would only see this if clone_2 at the beginning if d happened before c right?
Also if you don't sort at all, versus sorting per backward node, that also changes things...